File: src-tauri/src/models/terminal.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TerminalSession {
    pub id: String,
    pub workspace_id: String,
    pub session_role: String,
    pub profile: String,
    pub cwd: String,
    pub status: String,
    pub started_at: String,
    pub ended_at: Option<String>,
    pub command: String,
    pub args: Vec<String>,
    pub pid: Option<u32>,
    pub stale: bool,
    pub closed_at: Option<String>,
    pub backend: String,
    pub title: String,
    pub terminal_kind: String,
    pub display_order: i64,
    pub is_visible: bool,
    pub last_attached_at: Option<String>,
    pub last_captured_seq: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TerminalOutputChunk {
    pub id: String,
    pub session_id: String,
    pub seq: u64,
    pub timestamp: String,
    pub stream_type: String,
    pub data: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TerminalSessionState {
    pub active_session: Option<TerminalSession>,
    pub latest_session: Option<TerminalSession>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TerminalOutputResponse {
    pub session: Option<TerminalSession>,
    pub chunks: Vec<TerminalOutputChunk>,
    pub next_seq: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TerminalOutputEvent {
    pub workspace_id: String,
    pub chunk: TerminalOutputChunk,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CommandApprovalEvent {
    pub session_id: String,
    pub workspace_id: String,
    /// The human-readable command text (stripped of trailing newlines).
    pub command: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StartTerminalSessionInput {
    pub workspace_id: String,
    pub profile: String,
    pub session_role: Option<String>,
    pub cols: Option<u16>,
    pub rows: Option<u16>,
    pub replace_existing: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateWorkspaceTerminalInput {
    pub workspace_id: String,
    pub kind: String,
    pub profile: String,
    pub profile_id: Option<String>,
    pub title: Option<String>,
    pub command: Option<String>,
    pub args: Option<Vec<String>>,
    pub cols: Option<u16>,
    pub rows: Option<u16>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AttachWorkspaceTerminalInput {
    pub workspace_id: String,
    pub session_id: String,
    pub cols: Option<u16>,
    pub rows: Option<u16>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QueueAgentPromptInput {
    pub workspace_id: String,
    pub prompt: String,
    pub profile: Option<String>,
    pub profile_id: Option<String>,
    pub task_mode: Option<String>,
    pub reasoning: Option<String>,
    pub mode: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BatchDispatchPromptInput {
    pub workspace_ids: Vec<String>,
    pub prompt: String,
    pub profile_id: Option<String>,
    pub task_mode: Option<String>,
    pub reasoning: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentPromptEntry {
    pub id: String,
    pub workspace_id: String,
    pub session_id: Option<String>,
    pub profile: String,
    pub prompt: String,
    pub status: String,
    pub created_at: String,
    pub sent_at: Option<String>,
}

File: src/types/terminal.ts
export type TerminalProfile = 'shell' | 'codex' | 'claude_code' | 'kimi_code' | string;
export type TerminalSessionStatus = 'running' | 'succeeded' | 'failed' | 'stopped' | 'interrupted' | string;

export interface TerminalSession {
  id: string;
  workspaceId: string;
  sessionRole?: 'agent' | 'utility' | string;
  profile: TerminalProfile | string;
  cwd: string;
  status: TerminalSessionStatus;
  startedAt: string;
  endedAt?: string;
  command: string;
  args: string[];
  pid?: number;
  stale: boolean;
  closedAt?: string;
  backend: string;
  title: string;
  terminalKind: 'agent' | 'shell' | 'run' | 'utility' | string;
  displayOrder: number;
  isVisible: boolean;
  lastAttachedAt?: string;
  lastCapturedSeq: number;
}

export interface TerminalOutputChunk {
  id: string;
  sessionId: string;
  seq: number;
  timestamp: string;
  streamType: 'pty' | 'system' | string;
  data: string;
}

export interface TerminalOutputEvent {
  workspaceId: string;
  chunk: TerminalOutputChunk;
}

export interface TerminalSessionState {
  activeSession?: TerminalSession | null;
  latestSession?: TerminalSession | null;
}

export interface TerminalOutputResponse {
  session?: TerminalSession | null;
  chunks: TerminalOutputChunk[];
  nextSeq: number;
}

export interface StartTerminalSessionInput {
  workspaceId: string;
  profile: TerminalProfile;
  sessionRole?: 'agent' | 'utility';
  cols?: number;
  rows?: number;
  replaceExisting?: boolean;
}

export interface CreateWorkspaceTerminalInput {
  workspaceId: string;
  kind: 'agent' | 'shell' | 'run' | 'utility';
  profile: TerminalProfile;
  profileId?: string;
  title?: string;
  command?: string;
  args?: string[];
  cols?: number;
  rows?: number;
}

export interface AttachWorkspaceTerminalInput {
  workspaceId: string;
  sessionId: string;
  cols?: number;
  rows?: number;
}

export type AgentPromptStatus = 'queued' | 'sent' | 'succeeded' | 'failed' | 'interrupted' | 'stopped' | string;

export interface AgentPromptEntry {
  id: string;
  workspaceId: string;
  sessionId?: string;
  profile: TerminalProfile | string;
  prompt: string;
  status: AgentPromptStatus;
  createdAt: string;
  sentAt?: string;
}

export interface QueueAgentPromptInput {
  workspaceId: string;
  prompt: string;
  profile?: TerminalProfile;
  profileId?: string;
  taskMode?: string;
  reasoning?: string;
  /** Optional; server always dispatches to the agent terminal after recording the prompt. */
  mode?: 'send_now' | 'interrupt_send';
}

export interface TerminalSearchResult {
  workspaceId: string;
  workspaceName: string;
  sessionId: string;
  timestamp: string;
  line: string;
}
